README.md
## Lms

Loan Managment System

#### License

MIT


CHANGELOG.md
# Changelog

## [Unreleased]

## [v5.17.5] - 2023-06-08

## Added

- [Loan Is closed flag handling](https://gitlab.com/omkar.darves/choicelms/-/merge_requests/1)

## [v5.17.4] - 2023-05-24

## Fix

- [Restrict withdrawal if balance is > 0, topup and pledge for MF, mark loan as is_closed if total collateral,balance and sum of virtual interest not booked is 0](https://gitlab.com/atri-tech/atri-maintainers/choicelms/-/merge_requests/588)

## [v5.17.3] - 2023-05-22

## Added

- [Code cleaning and KYC restrict](https://gitlab.com/atri-tech/atri-maintainers/choicelms/-/merge_requests/587)

## [v5.17.2] - 2023-05-18

## Fix

- [Collateral legder not updating](https://gitlab.com/atri-tech/atri-maintainers/choicelms/-/merge_requests/586)

## [v5.17.1] - 2023-05-10

## Added

- [SPRK-43 - Offline customer traceback fixed and Custom Roi slab fix](https://gitlab.com/atri-tech/atri-maintainers/choicelms/-/merge_requests/585)

## [v5.17.0] - 2023-05-03

## Added

- [SPRK-43 - Increase Loan In MF](https://gitlab.com/atri-tech/atri-maintainers/choicelms/-/merge_requests/584)

## [v5.16.0] - 2023-04-11

## Added

- [SPRK-48 - Loan Renewal and Custom ROI](https://gitlab.com/atri-tech/atri-maintainers/choicelms/-/merge_requests/582)

## [Unreleased]

## [v5.15.0] - 2023-03-30

## Added

- [SPRK-128 - Ltv for Shares and Shares Eligibility](https://gitlab.com/atri-tech/atri-maintainers/choicelms/-/merge_requests/581)

## [v5.14.1] - 2023-03-23

## Added

- [SPRK-122 - Read Only fields updated in loan application, pan no and dob in user kyc doctype mark read only only on prod](https://gitlab.com/atri-tech/atri-maintainers/choicelms/-/merge_requests/578)

## [v5.14.0] - 2023-03-13

## Added

- [SPRK-122 - Offline Loan, Au penny name matching, Optional Update for mobile app](https://gitlab.com/atri-tech/atri-maintainers/choicelms/-/merge_requests/576)

## [v5.13.8] - 2023-03-06

## Added

- [SPRK 176 - Code and Cron optimization - Process Pledge cron(CDSL pledge setup)](https://gitlab.com/atri-tech/atri-maintainers/choicelms/-/merge_requests/574)

## [v5.13.7] - 2023-03-03

## Added

- [SPRK-156 - Update Razorpay webhook function and increase background workers for cron jobs execution](https://gitlab.com/atri-tech/atri-maintainers/choicelms/-/merge_requests/573)

## [v5.13.6] - 2023-02-28

## Added

- [SPRK-60/SPRK-107 Pledge Application Creation/Workflow state change issue](https://gitlab.com/atri-tech/atri-maintainers/choicelms/-/commit/8a00079de52bfd903d5641398a58f32c34728369)

## [v5.13.5] - 2023-02-27

## Added

- [SPRK-144 - if last name or middle name field is empty then 1st name to be considered as full for name matching with penny retrieved name](https://gitlab.com/atri-tech/atri-maintainers/choicelms/-/commit/ecb49130440c7872ccdece26041c513b2f14558a)

## [v5.13.4] - 2023-02-22

## Fixed

- [Feat : Sanction letter and onboarding Screen](https://gitlab.com/atri-tech/atri-maintainers/choicelms/-/merge_requests/572)
- [Feat : Onboarding screen changes](https://gitlab.com/atri-tech/atri-maintainers/choicelms/-/merge_requests/571)

## [v5.13.3] - 2023-02-13

## Fixed

- [Feat : Onboarding screen changes](https://gitlab.com/atri-tech/atri-maintainers/choicelms/-/merge_requests/571)

## [v5.13.2] - 2023-01-23

## Fixed

- [Fix : Increase loan fix and security exposure fix](https://gitlab.com/atri-tech/atri-maintainers/choicelms/-/merge_requests/569)

## [v5.13.1] - 2023-01-23

## Fixed

- [Fix : Website bug fix prod](https://gitlab.com/atri-tech/atri-maintainers/choicelms/-/merge_requests/567)

## [v5.13.0] - 2023-01-19

## Added

- [Merge branch Uat into Master](https://gitlab.com/atri-tech/atri-maintainers/choicelms/-/merge_requests/566)

## [v5.13.0-uat]

## Added

- [Feat: Update security and allowed security patches](https://gitlab.com/atri-tech/atri-maintainers/choicelms/-/merge_requests/565)

## [v5.12.0-uat]

## Added

- [Feat: Annexure update in privacy policy update](https://gitlab.com/atri-tech/atri-maintainers/choicelms/-/merge_requests/564)

## [v5.11.0-uat]

## Added

- [Feat: Choice office Address change on website and au pennydrop add bank before 90 days changes](https://gitlab.com/atri-tech/atri-maintainers/choicelms/-/merge_requests/563)

## [v5.10.3-uat]

## Fixed

- [Fix: unpldege and sell collateral psn handling](https://gitlab.com/atri-tech/atri-maintainers/choicelms/-/merge_requests/561)

## [v5.10.2-uat]

## Added

- [Feat: floating-to-fixed](https://gitlab.com/atri-tech/atri-maintainers/choicelms/-/merge_requests/560)

## [v5.10.1-uat]

## Added

- [Feat: floating-to-fixed](https://gitlab.com/atri-tech/atri-maintainers/choicelms/-/merge_requests/559)

## [v5.10.0-uat]

## Added

- [Feat: live penny integration with fullname matching](https://gitlab.com/atri-tech/atri-maintainers/choicelms/-/merge_requests/557)

## [v5.9.0+1-uat]

## Added

- [Feat: penny drop hit with same account number](https://gitlab.com/atri-tech/atri-maintainers/choicelms/-/merge_requests/554)

## [v5.9.0-uat]

## Added

- [Feat: AU Bank Penny Drop Api integration](https://gitlab.com/atri-tech/atri-maintainers/choicelms/-/merge_requests/552)

## [v5.8.2-uat]

## Added

- [Feat: \n added in unpledge application msg](https://gitlab.com/atri-tech/atri-maintainers/choicelms/-/merge_requests/547)

## [v5.8.1-uat]

## Added

- [Feat: update dlt links patch changes](https://gitlab.com/atri-tech/atri-maintainers/choicelms/-/merge_requests/546)

## [v5.8.0-uat]

## Added

- [Feat: DLT sms changes and 4 fields added in las settings for dlt](https://gitlab.com/atri-tech/atri-maintainers/choicelms/-/merge_requests/544)

## [v5.7.2-uat]

## Fixed

- [Fix: Margin Shortfall Amount update in Loan](https://gitlab.com/atri-tech/atri-maintainers/choicelms/-/merge_requests/542)

## [v5.7.1-uat]

## Fixed

- [Fix: Address in get tnc fix](https://gitlab.com/atri-tech/atri-maintainers/choicelms/-/merge_requests/540)

## [v5.7.0-uat]

## Added


lms/hooks.py
# -*- coding: utf-8 -*-
from __future__ import unicode_literals

from . import __version__ as app_version

app_name = "lms"
app_title = "Lms"
app_publisher = "Atrina Technologies Pvt. Ltd."
app_description = "Loan Managment System"
app_icon = "octicon octicon-file-directory"
app_color = "grey"
app_email = "developers@atritechnocrat.com"
app_license = "MIT"
app_logo_url = "/assets/lms/images/logo_mo.svg"

# Includes in <head>
# ------------------

# include js, css files in header of desk.html
# app_include_css = "/assets/lms/css/lms.css"
# app_include_js = "/assets/lms/js/lms.js"
app_include_js = "/assets/lms/js/doctype.js"

# include js, css files in header of web template
# web_include_css = "/assets/lms/css/lms.css"
# web_include_js = "/assets/lms/js/lms.js"

# include js in page
# page_js = {"page" : "public/js/file.js"}
# include js in doctype views
# doctype_js = {"doctype" : "public/js/doctype.js"}
# doctype_list_js = {"doctype" : "public/js/doctype_list.js"}
# doctype_tree_js = {"doctype" : "public/js/doctype_tree.js"}
# doctype_calendar_js = {"doctype" : "public/js/doctype_calendar.js"}

# Home Pages
# ----------

# application home page (will override Website Settings)
home_page = "home"

# website user home page (by Role)
# role_home_page = {
# 	"Role": "home_page"
# }

# Website user home page (by function)
# get_website_user_home_page = "lms.utils.get_home_page"

# Generators
# ----------

# automatically create page for each record of this doctype
# website_generators = ["Web Page"]

# Installation
# ------------

# before_install = "lms.install.before_install"
after_install = "lms.after_install"

# Desk Notifications
# ------------------
# See frappe.core.notifications.get_notification_config

# notification_config = "lms.notifications.get_notification_config"

# Permissions
# -----------
# Permissions evaluated in scripted ways

permission_query_conditions = {
    # "Event": "frappe.desk.doctype.event.event.get_permission_query_conditions",
    # "Lender": "lms.hooks.lender_permission_query",
    "Cart": "lms.cart_permission_query",
    "Loan Application": "lms.loan_application_permission_query",
    "Collateral Ledger": "lms.collateral_ledger_permission_query",
    "Loan": "lms.loan_permission_query",
    "Loan Transaction": "lms.loan_transaction_permission_query",
    "Loan Margin Shortfall": "lms.loan_margin_shortfall_permission_query",
    "Unpledge Application": "lms.unpledge_application_permission_query",
    "Sell Collateral Application": "lms.sell_collateral_application_permission_query",
    "Top up Application": "lms.top_up_application_permission_query",
    "Virtual Interest": "lms.virtual_interest_permission_query",
    "Lender Ledger": "lms.lender_ledger_permission_query",
    "Allowed Security": "lms.allowed_security_permission_query",
    "Interest Configuration": "lms.interest_configuration_permission_query",
    "Lender": "lms.lender_permission_query",
    "Loan Payment Log": "lms.loan_payment_log_permission_query",
    "Security Category": "lms.security_category_permission_query",
}

#
# has_permission = {
# 	"Event": "frappe.desk.doctype.event.event.has_permission",
# }

# Document Events
# ---------------
# Hook on document methods and events

doc_events = {
    "User": {"on_trash": "lms.delete_user"},
    "File": {
        "before_insert": [
            "lms.lms.doctype.loan_application.loan_application.only_pdf_upload",
            "lms.lms.doctype.top_up_application.top_up_application.only_pdf_upload",
        ]
    },
}

# Scheduled Tasks
# ---------------

# scheduler_events = {
# 	"all": [
# 		"lms.tasks.all"
# 	],
# 	"daily": [
# 		"lms.tasks.daily"
# 	],
# 	"hourly": [
# 		"lms.tasks.hourly"
# 	],
# 	"weekly": [
# 		"lms.tasks.weekly"
# 	]
# 	"monthly": [
# 		"lms.tasks.monthly"
# 	]
# }

# Testing
# -------

# before_tests = "lms.install.before_tests"

# Overriding Methods
# ------------------------------
#
# override_whitelisted_methods = {
# 	"frappe.desk.doctype.event.event.get_events": "lms.event.get_events"
# }
#
# each overriding function accepts a `data` argument;
# generated from the base implementation of the doctype dashboard,
# along with any modifications made in other Frappe apps
override_doctype_dashboards = {
    # "Task": "lms.task.get_dashboard_data"
    "User": "lms.user_dashboard"
}

fixtures = [
    "API Doc",
]

scheduler_events = {
    "hourly": [
        "lms.lms.doctype.security_price.security_price.update_all_security_prices"
    ],
    # "daily": [
    #     "lms.lms.doctype.loan.loan.add_all_loans_virtual_interest",
    #     "lms.lms.doctype.loan.loan.check_for_all_loans_additional_interest",
    #     "lms.lms.doctype.loan.loan.add_all_loans_penal_interest",
    # ],
    # "monthly": ["lms.lms.doctype.loan.loan.book_all_loans_virtual_interest_for_month"],
    # "all": ["lms.lms.doctype.loan_margin_shortfall.loan_margin_shortfall.mark_sell_triggered"],
    "cron": {
        "*/5 * * * *": [
            "lms.lms.doctype.loan_application.loan_application.process_pledge",
            "lms.lms.doctype.loan_margin_shortfall.loan_margin_shortfall.mark_sell_triggered",
        ],  # At every 5 minutes
        "30 17,5 * * *": [
            "lms.lms.doctype.loan_transaction.loan_transaction.reject_blank_transaction_and_settlement_recon_api"
        ],
        "0 0 * * *": [
            "lms.lms.doctype.loan.loan.add_all_loans_virtual_interest"
        ],  # At 12:00 AM daily
        "0 1 * * *": [
            "lms.lms.doctype.loan.loan.book_all_loans_virtual_interest_for_month"
        ],  # At 01:00 AM on 1st day-of-every-month(monthly)
        "0 2 * * *": [
            "lms.lms.doctype.loan.loan.check_for_all_loans_additional_interest"
        ],  # At 02:00 AM daily
        "0 3 * * *": [
            "lms.lms.doctype.loan.loan.add_all_loans_penal_interest"
        ],  # At 03:00 AM daily
        "30 9 * * *": [
            "lms.lms.doctype.security_price.security_price.update_all_schemeNav"
        ],  # At 09:30 AM daily
        # "* * * * *": [
        #     "lms.lms.doctype.loan_margin_shortfall.loan_margin_shortfall.mark_sell_triggered"
        # ],  # At every minute
        "30 23 * * *": ["lms.system_report_enqueue"],  # At 11:30 PM daily
        "15 0 * * *": [
            "lms.lms.doctype.spark_loan_renewal_application.spark_loan_renewal_application.all_loans_renewal_update_doc"
        ],  # At 12:15 AM daily
    },
}
